### ISSUE TITLE
Nested path comprehension results to 'None.get' 

### ISSUE THREAD

**Issue by ischerbakov** [neo4j/neo4j#9377]
Nested path comprehension results to 'None.get' `Neo.DatabaseError.General.UnknownError`

**Neo4j version**: 3.2.0 (also reproduced in 3.1.2 but not in 3.1.1)
**Operating system**: Windows Server 2012 R2

The query:
```
MATCH (root{id:"root"})
RETURN {
   foo: [(root)-[:REL1]->(n) | {
      type: [(n)-[:TYPE]->(t) | t][0]
   }][0],
   bar: [(root)-[:REL2]->(n) | {
      type: [(n)-[:TYPE]->(t) | t][0]
   }][0]
} AS root
```
But the query below (when removed one 2nd lvl comprehension) works as expected:
```
MATCH (root{id:"root"})
RETURN {
   foo: [(root)-[:REL1]->(n) | {
   }][0],
   bar: [(root)-[:REL2]->(n) | {
      type: [(n)-[:TYPE]->(t) | t][0]
   }][0]
} AS root
```
**bebug.log:**
```
ERROR [o.n.b.v.r.ErrorReporter] Client triggered an unexpected error [UnknownError]: None.get, reference c65c3f2e-879c-4095-96bf-1f9b12ef7519. None.get
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:347)
	at scala.None$.get(Option.scala:345)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.Expression.owningPipe(Expression.scala:58)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.NestedPipeExpression.apply(NestedPipeExpression.scala:31)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.NullInNullOutExpression.apply(NullInNullOutExpression.scala:28)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.LiteralMap$$anonfun$apply$1.apply(LiteralMap.scala:31)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.LiteralMap$$anonfun$apply$1.apply(LiteralMap.scala:30)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.LiteralMap.apply(LiteralMap.scala:30)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.NestedPipeExpression$$anonfun$apply$1.apply(NestedPipeExpression.scala:32)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.NestedPipeExpression$$anonfun$apply$1.apply(NestedPipeExpression.scala:32)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:59)
	at scala.collection.immutable.VectorBuilder.$plus$plus$eq(Vector.scala:732)
	at scala.collection.immutable.VectorBuilder.$plus$plus$eq(Vector.scala:708)
	at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:310)
	at scala.collection.AbstractIterator.to(Iterator.scala:1336)
	at scala.collection.TraversableOnce$class.toIndexedSeq(TraversableOnce.scala:300)
	at scala.collection.AbstractIterator.toIndexedSeq(Iterator.scala:1336)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.NestedPipeExpression.apply(NestedPipeExpression.scala:32)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.NullInNullOutExpression.apply(NullInNullOutExpression.scala:28)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.LiteralMap$$anonfun$apply$1.apply(LiteralMap.scala:31)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.LiteralMap$$anonfun$apply$1.apply(LiteralMap.scala:30)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.Map$Map2.foreach(Map.scala:137)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.neo4j.cypher.internal.compiler.v3_2.commands.expressions.LiteralMap.apply(LiteralMap.scala:30)
	at org.neo4j.cypher.internal.compiler.v3_2.pipes.ProjectionPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(ProjectionPipe.scala:41)
	at org.neo4j.cypher.internal.compiler.v3_2.pipes.ProjectionPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(ProjectionPipe.scala:39)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
	at org.neo4j.cypher.internal.compiler.v3_2.pipes.ProjectionPipe$$anonfun$internalCreateResults$1.apply(ProjectionPipe.scala:39)
	at org.neo4j.cypher.internal.compiler.v3_2.pipes.ProjectionPipe$$anonfun$internalCreateResults$1.apply(ProjectionPipe.scala:38)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator$$anonfun$next$1.apply(ResultIterator.scala:71)
	at org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator$$anonfun$next$1.apply(ResultIterator.scala:68)
	at org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator$$anonfun$failIfThrows$1.apply(ResultIterator.scala:94)
	at org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator.decoratedCypherException(ResultIterator.scala:103)
	at org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator.failIfThrows(ResultIterator.scala:92)
	at org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator.next(ResultIterator.scala:68)
	at org.neo4j.cypher.internal.compiler.v3_2.ClosingIterator.next(ResultIterator.scala:49)
	at org.neo4j.cypher.internal.compiler.v3_2.PipeExecutionResult.next(PipeExecutionResult.scala:79)
	at org.neo4j.cypher.internal.compiler.v3_2.PipeExecutionResult.next(PipeExecutionResult.scala:37)
	at org.neo4j.cypher.internal.compiler.v3_2.helpers.RuntimeJavaValueConverter$feedIteratorToVisitable.accept(RuntimeJavaValueConverter.scala:54)
	at org.neo4j.cypher.internal.compiler.v3_2.PipeExecutionResult.accept(PipeExecutionResult.scala:109)
	at org.neo4j.cypher.internal.compatibility.v3_2.ExecutionResultWrapper.accept(ExecutionResultWrapper.scala:159)
	at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$accept$1.apply$mcV$sp(ClosingExecutionResult.scala:169)
	at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$accept$1.apply(ClosingExecutionResult.scala:168)
	at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult$$anonfun$accept$1.apply(ClosingExecutionResult.scala:168)
	at org.neo4j.cypher.internal.compatibility.v3_2.exceptionHandler$runSafely$.apply(exceptionHandler.scala:90)
	at org.neo4j.cypher.internal.compatibility.ClosingExecutionResult.accept(ClosingExecutionResult.scala:168)
	at org.neo4j.cypher.internal.javacompat.ExecutionResult.accept(ExecutionResult.java:318)
	at org.neo4j.bolt.v1.runtime.CypherAdapterStream.accept(CypherAdapterStream.java:69)
	at org.neo4j.bolt.v1.messaging.BoltMessageRouter$ResultHandler.onRecords(BoltMessageRouter.java:130)
	at org.neo4j.bolt.v1.runtime.BoltStateMachine$State$3.lambda$pullAll$0(BoltStateMachine.java:448)
	at org.neo4j.bolt.v1.runtime.TransactionStateMachine$State$1.streamResult(TransactionStateMachine.java:227)
	at org.neo4j.bolt.v1.runtime.TransactionStateMachine.streamResult(TransactionStateMachine.java:90)
	at org.neo4j.bolt.v1.runtime.BoltStateMachine$State$3.pullAll(BoltStateMachine.java:447)
	at org.neo4j.bolt.v1.runtime.BoltStateMachine.pullAll(BoltStateMachine.java:229)
	at org.neo4j.bolt.v1.messaging.BoltMessageRouter.lambda$onPullAll$6(BoltMessageRouter.java:98)
	at org.neo4j.bolt.v1.runtime.concurrent.RunnableBoltWorker.execute(RunnableBoltWorker.java:130)
	at org.neo4j.bolt.v1.runtime.concurrent.RunnableBoltWorker.executeBatch(RunnableBoltWorker.java:123)
	at org.neo4j.bolt.v1.runtime.concurrent.RunnableBoltWorker.run(RunnableBoltWorker.java:96)
	at java.lang.Thread.run(Unknown Source)
```

**Comment by ischerbakov** [comment]
How can I get around this to get the same shape as a result?
I'm in hurry, and my project relies on pattern comprehension almost completely (it's a GraphQL server).

**Comment by craigtaverner** [comment]
We have a seen a similar error in another case and perhaps the fix for that will apply to your case. However, I cannot reproduce your error in 3.1.1. I've created sample data using:

    CREATE (root{id:"root"})
    CREATE (root)-[:REL1]->(n1)-[:TYPE]->(t1)
    CREATE (root)-[:REL2]->(n2)-[:TYPE]->(t2)

And then run the query:

    MATCH (root{id:"root"})
    RETURN {
       foo: [(root)-[:REL1]->(n) | {
          type: [(n)-[:TYPE]->(t) | t][0]
       }][0],
       bar: [(root)-[:REL2]->(n) | {
          type: [(n)-[:TYPE]->(t) | t][0]
       }][0]
    } AS x

And this works. Could you tell us more about your case so we can try reproduce it? How much data, what indexes, query profile? For listing the indexes:

    :schema

For getting the query plan, just prefix your query with `PROFILE`.

One thing to try to get around the issue is to move the nested comprehensions out of the map literal. For example, try:

    MATCH (root{id:"root"})
    WITH [(root)-[:REL1]->(n) | {
          type: [(n)-[:TYPE]->(t) | t][0]
       }] as foo,
       [(root)-[:REL2]->(n) | {
          type: [(n)-[:TYPE]->(t) | t][0]
       }] as bar
    RETURN {
       foo: foo[0],
       bar: bar[0]
    } AS x

**Comment by craigtaverner** [comment]
Update: I was able to reproduce this in 3.2.0. My initial tests above were in 3.1.1, where the bug does not appear to happen. So it seems to occur in later versions of 3.1.x, and in 3.2.0.

The workaround above does work for 3.2.0.

**Comment by ischerbakov** [comment]
There's about 100 nodes and about 200 relationships in my database, no indexes, no constraints.
`PROFILE` fails with the same error.
I can not reproduce the error in 3.1.1, too. But I've just reproduced it in 3.1.2.

Since I have my data from 3.1.2, I've decided to work with 3.1.1 for a while.
Anyway, thank you for the workaround! It does work for 3.1.2, too.

**Comment by fickludd** [comment]
Thank you for reporting @ischerbakov,

This issue will be solved by #9389, which will likely make 3.1.5 and 3.2.1.
